src/lib/tls/asio/asio_stream.h

```diff
@@ -0,0 +1,585 @@
+       */
+      template <typename ConstBufferSequence, typename WriteHandler>
+      BOOST_ASIO_INITFN_RESULT_TYPE(WriteHandler,
+                                    void(boost::system::error_code, std::size_t))
+      async_write_some(const ConstBufferSequence& buffers, WriteHandler&& handler)
+         {
+         BOOST_ASIO_WRITE_HANDLER_CHECK(WriteHandler, handler) type_check;
+
+         boost::asio::async_completion<WriteHandler, void(boost::system::error_code, std::size_t)> init(handler);
+
+         boost::system::error_code ec;
+         std::size_t sent = tls_encrypt_some(buffers, ec);
+         if(ec)
+            {
+            // we cannot be sure how many bytes were committed here
+            // so clear the send_buffer and let the AsyncWriteOperation call the handler with the error_code set
+            this->m_core.clearSendBuffer();
+            Botan::TLS::AsyncWriteOperation<typename std::decay<WriteHandler>::type, Stream>
+            op{std::move(init.completion_handler), *this, this->m_core, std::size_t(0), ec};
+            return init.result.get();
+            }
+
+         Botan::TLS::AsyncWriteOperation<typename std::decay<WriteHandler>::type, Stream>
+         op{std::move(init.completion_handler), *this, this->m_core, sent};
+
+         return init.result.get();
+         }
+
+      /**
+       * Start an asynchronous read. The function call always returns immediately.
+       *
+       * @param buffers The buffers into which the data will be read. Although the buffers object may be copied as
+       *                necessary, ownership of the underlying buffers is retained by the caller, which must guarantee
+       *                that they remain valid until the handler is called.
+       * @param handler The handler to be called when the read operation completes. The equivalent function signature of
+       *                the handler must be:
+       *                void(boost::system::error_code, std::size_t)
+       */
+      template <typename MutableBufferSequence, typename ReadHandler>
+      BOOST_ASIO_INITFN_RESULT_TYPE(ReadHandler,
+                                    void(boost::system::error_code, std::size_t))
+      async_read_some(const MutableBufferSequence& buffers, ReadHandler&& handler)
+         {
+         BOOST_ASIO_READ_HANDLER_CHECK(ReadHandler, handler) type_check;
+
+         boost::asio::async_completion<ReadHandler, void(boost::system::error_code, std::size_t)> init(handler);
+
+         AsyncReadOperation<typename std::decay<ReadHandler>::type, Stream, MutableBufferSequence>
+         op{std::move(init.completion_handler), *this, this->m_core, buffers};
+         return init.result.get();
+         }
+
+   protected:
+      size_t writePendingTlsData(boost::system::error_code& ec)
+         {
+         auto writtenBytes = boost::asio::write(m_nextLayer, this->m_core.sendBuffer(), ec);
+
+         this->m_core.consumeSendBuffer(writtenBytes);
```

Comment:
For my understanding of the buffer handling in the `StreamCore`. Imagine the following:

* application calls the synchronous `write_some()` several times with the biggest possible buffer
* `tls_encrypt_some` encrypts the data (probably causing some overhead) and stores it in the send buffer
* `asio::write()` pushes "some" data out through the socket, but leaves some data behind in the send buffer
* since `write_some()` calls `writePendingTlsData()` only once after `tls_encrypt_data`
  * shouldn't we see some data build-up in the send buffer that could eventually cause a dead lock or data not being sent as quickly as possible or being dropped?

Is this taken care of? and how?